Refactor notification init

bebehei opened the pull request:
Refactors `notification_init` to split it into multiple methods.

tsipinakis commented on src/notification.c:
```diff
@@ -280,55 +283,96 @@ char *notification_extract_markup_urls(char **str_ptr) {
 {
-        return g_malloc0(sizeof(notification));
-}
+        //FIXME: handle OOM
```
From [glibs docs](https://developer.gnome.org/glib/stable/glib-Memory-Allocation.html)
> If any call to allocate memory fails, the application is terminated. This also means that there is no need to check if the call succeeded.

Means we don't need to handle OOM.

bebehei replied:
Thanks. This wasn't clear for me. Only [g_try_malloc0](https://developer.gnome.org/glib/stable/glib-Memory-Allocation.html#g-try-malloc0) mentions the behavior of `g_malloc0`.

tsipinakis commented on src/notification.c:
```diff
@@ -280,55 +283,96 @@ char *notification_extract_markup_urls(char **str_ptr) {
+        /* Color hints */
+        n->colors[ColFG] = n->colors[ColFG]
+                         ? n->colors[ColFG]
+                         : xctx.colors[ColFG][n->urgency];
```
I *really* don't like using `xctx` inside non-X related functions. Though I guess this will have to stay as-is until the drawing refactor is done.

bebehei replied:
I agree with you. But previously, `notification_init` also had to access this struct. It's bad, but it's not getting worse (and hopefully gets fixed in refactor drawing).

tsipinakis replied:
Also, I can't help but feel like all these ternary operators are a bit ugly - what's wrong with a usual `if`?

tsipinakis commented on src/notification.c:
```diff
@@ -429,41 +473,25 @@ void notification_init(notification *n)
+static void notification_extract_urls(notification *n)
+{
+        // DO markup urls processing here until we split this out correctly
+        n->urls = notification_extract_markup_urls(&(n->body)); //WHAT THE FUCK!?
```
I'm not sure what the `WTF` is here for? - Sure not the best code but not really weird?

bebehei replied:
Thank you for mentioning this again.

For me, this function call is against every logic. Why do we put a double pointer to that (and not the notification and turn it into a method like every other method).

I also don't see, why we do have to make changes to the body field there. Markup should be extracted somewhere else.

bebehei replied:
1. There is a discrepancy between pango and notification spec. Notification spec requires `img` and `a` tags, while pango cannot render these (obviously). As `a` tags happen quite regularly, these have to get replaced. Guess, what this method did, too!?

2. Let's have these two notifications:
    - `notify-send asdf '<a href="https://github.com/">testname</a>'`
        - ` n->urls = 0x5555558675f0 "[#1] https://github.com/"`
    - `notify-send https://summary-url.de https://internal.com`
        - `1: n->urls = 0x555555803980 "https://summary-url.de\nhttps://internal.com"`
    - Mhmm, same field, different format, 👎

bebehei commented:
@tsipinakis As I'm tired right now and may have introduced some bugs by myself with the lastest commits, I will review this by myself at first. I will "ping" you by re-requesting a review.

---

OT: I would be able to click on merge for my own PRs, while you requested changes in your current review. Would you mind changing this to make a positive review a neccessity?

tsipinakis commented:
> OT: I would be able to click on merge for my own PRs, while you  requested changes in your current review. Would you mind changing this  to make a positive review a neccessity?

I tweaked the branch restrictions a bit and set it to require a positive review and a passing build to push to master. Thought technically you'd be able to review and approve your own commit, right?

bebehei commented:
> Thought technically you'd be able to review and approve your own pr, right?

I guess not. Let's test it!

bebehei commented:
@tsipinakis has enlightened me about the previous functionality of `notification_extract_markup_urls` right now in IRC.

I heavily want to drop this ugly function and replace it with something similar. The specific functionality of adding `[ #linkno]` in front of the URL is quite bogus for me. I don't see why this is neccessary. IMHO, there is `show_indicators` for such functionality. And also having the linknumber included in the urls field makes it quite ugly, as there is no consistency to `extract_urls`.

Currently I have also removed it in this PR and I'd like to evaluate which path we want to go.

bebehei commented:
> [...] and I'd like to evaluate which path we want to go.

@tsipinakis Could you please give your opinion here?

tsipinakis commented on src/notification.h:
```diff
@@ -30,34 +30,44 @@ typedef struct _actions {
 } Actions;
 
 typedef struct _notification {
+        int id;
```
I'd group this differently: `id` and `dbus_client` together and `appname`, `summary`, `body`, `category`, `urgency` in another group since they're the main notification attributes.

bebehei replied:
Grouped it accordingly now.

tsipinakis commented:
> The specific functionality of adding [ #linkno] in front of the URL is quite bogus for me. I don't see why this is necessary.

As I've explained on IRC, as I view it this is because there is no clear indication of which url belongs to which piece of text in the notification when it comes to `a href` tags. For plain urls this is not necessary since it's displayed as text anyway.

> IMHO, there is show_indicators for such functionality.

`show_indicators` shows that there's a url in the notification - it doesn't provide any information as to which piece of text contains which url.

> (Edit: I also found, that notification spec mentions to show it in standard blue underlined text.)

We should probably comply with this but it might be a bit misleading since clicking on it won't work with the current implementation.

Currently I have also removed it in this PR and I'd like to evaluate which path we want to go.

bebehei commented:
> We should probably comply with this but it might be a bit misleading since clicking on it won't work with the current implementation.

Question to understand this: Does pango support $something to know, that the user clicked on a specific textbit?

tsipinakis commented:
> Does pango support $something to know, that the user clicked on a specific textbit?

Quick google didn't turn up anything - I have no clue.

bebehei commented:
> it doesn't provide any information as to which piece of text contains which url.

what's better than a linknumber? Put in the actual text! So, if you open up dmenu it will display `[linktext] URL`.

e.g:
```
notify-send '<a href="https://github.com/dunst-project/dunst/pull/428">GitHub</a>' 'Pull Request updated'
```

will result in:

```
n->urls = "[GitHub] https://github.com/dunst-project/dunst/pull/428"
```

This is also more comfortable.

tsipinakis commented:
> what's better than a linknumber? Put in the actual text! So, if you open up dmenu it will display [linktext] URL.

While it seems like a good suggestion it could break in some scenarios like having a super long link text, or having two links with the same text

> The first example is _this_ and the second is _this_, which one do you prefer?

would produlce

> [this] URL     [this] URL

While the order would give it away somewhat, it's still confusing.

bebehei commented:
> [...] or having two links with the same text [...] While the order would give it away somewhat, it's still confusing.

I agree with you. It's actually confusing. But is it dunst's job to make confusing URLs clear? There is no information lost on dunst's behalf.

> While it seems like a good suggestion it could break in some scenarios like having a super long link text, 

Is a long URL currently handled better?

```
[ Hi Mike, I've got you a super long URL,
which renders faulty on your system but
it's good! Check it out! #1]
```

- How long do you need to parse that the URL number 1 takes the whole notification?
- What's the probability of having a second URL in a notification with a superlong URL?

---

I have to admit, I have only searched for a replacement of the current URL replacement, but I think the solution I came up with is actually much better than the current approach.

tsipinakis commented on src/menu.c:
```diff
@@ -95,8 +95,18 @@ char *extract_urls(const char *to_match)
-        // remove prefix and test url
-        char *url = extract_urls(in);
+
+        // remove leading [ linktext ] from URL
```
If we're removing the prefix here what's the point of still running it through `extract_urls`?

bebehei replied:
I thought I can catch some weird corner case with this. But after thinking again about it, I've proved myself wrong in this.

tsipinakis commented on src/markup.c:
```diff
@@ -64,7 +64,7 @@ static char *markup_strip_a(char *str)
                 unsigned int len = end ? (end-start)+1 : strlen(start);
 
-                str = string_replace_at(str, start-str, len, "");
+                str = string_replace_at(str, start-str, len, "<span color=\"blue\"><u>");
```
While it's specified by the spec, I can already see all the customization requests a mile away, especially since it looks ugly even with the default theme.

![2017-11-29t18-31-03](https://user-images.githubusercontent.com/18189112/33386364-7a00aac0-d533-11e7-8d45-404d9466765e.png)

bebehei replied:
Yes, this *is* ugly. I just pushed it as an example how the ugly the recommendation will look like, when we discussed, what the replacement for the link replacement method.

What would you like to do? Remove it fully and mention it to use the indicators? Make it more sensible by just underlining it?

tsipinakis replied:
I would say remove it for now and file an issue about it to look into it later.

tsipinakis commented on src/markup.c:
```diff
@@ -44,6 +45,66 @@ static char *markup_br2nl(char *str)
+
+        while ((start = strstr(start, "<img"))) {
+                end = strstr(start, ">");
+                alt = strstr(start, "alt=\""); // use alt=" as stated in the notification spec
```
How about handling `src` as a URL too?

bebehei replied:
Please, no. Wait for the feature request.

tsipinakis commented on src/notification.c:
```diff
@@ -438,45 +498,31 @@ void notification_init(notification *n)
+        char *concat = g_strconcat(n->summary, " ", n->body, NULL);
+        // remove <a href=... to not confuse extract_urls
+        char *concat_plain = markup_transform(g_strdup(concat), MARKUP_FULL);
+        char *urls_plain;
```
What's the point of this variable? You're assigning and freeing it.

bebehei replied:
It **is** actually used. But you're objection has a good reason. The way this variable gets assigned is shit.

tsipinakis commented on src/notification.c:
```diff
@@ -248,95 +255,148 @@ void notification_replace_single_field(char **haystack,
+
+                        const char *quote = strstr(href, "\"");
+
+                            if (quote && quote < tag1_end) {
```
What's up with the indentation here?

bebehei replied:
When you trust vim....

tsipinakis commented:
> Is a long URL currently handled better?
> [...]

Good point, is the time it takes to implement it and the switching costs worth it though? What are the upsides?

bebehei commented on src/markup.c:
```diff
@@ -44,6 +45,66 @@ static char *markup_br2nl(char *str)
+                end = strstr(start, ">");
+
+                // If we hit the end of the string, delete the trailing tag fully
+                unsigned int len = end ? (end-start)+1 : strlen(start);
```
Yes.

bebehei commented:
> I feel like this PR has grown from a simple refactor to way beyond scope changes -.-

Simple refactoring of `notification_init` is a first class oxymoron. 😂 

> Good point, is the time it takes to implement it and the switching costs worth it though? What are the upsides?

?

The upsides should be clear and the work is already done!?

Honestly, I don't get what you want to tell here.

tsipinakis commented:
> The upsides should be clear and the work is already done!?

What I was trying to say is "there is no reason to put in effort to fix what already works" but if it's already implemented I'm fine with merging it.

bebehei commented:
@tsipinakis  I have reworked the URL handling. It gets now done inside of `markup_strip_(a|img)`. I also added the behavior of adding image URLs. And it gets done over `n->msg` (the shown part of the notification).

But: This is all on my laptop. I won't commit it today. Tiredly reassembling the history is no good.

bebehei commented:
@tsipinakis I fixed all issues with the URL extraction methods. I added pelnty of tests. Have fun with "Remove a and img tags from msg". ;-)

bebehei commented on src/notification.c:
```diff
@@ -438,45 +475,25 @@ void notification_init(notification *n)
+{
         if (n->actions) {
-                n->actions->dmenu_str = NULL;
+                g_clear_pointer(&n->actions->dmenu_str, g_free);
```
There is no guard against calling this function a second time. Setting `dmenu_str` solely to `NULL` is irresponsible and makes absolutely no sense.

I've done this in every function, which rewrites some fields.

bebehei commented on src/markup.c:
```diff
@@ -44,6 +46,189 @@ static char *markup_br2nl(char *str)
+                }
+
+                // text between a tags
+                int text_len = tag2
```
You marked the exact line to let this code look as confusing as possible. 😂 

I fully agree with you.

bebehei commented on src/markup.c:
```diff
@@ -44,6 +46,189 @@ static char *markup_br2nl(char *str)
+                        src_e = strstr(src_s + strlen("src=\""), "\"");
+
+                // Move pointer to the actual start
+                alt_s = alt_s ? alt_s+5 : NULL;
```
Oh, yes. That's right.

tsipinakis commented on src/markup.c:
```diff
@@ -44,6 +46,189 @@ static char *markup_br2nl(char *str)
+                        // invisible in the replaced text
+                        if (!text_alt || strlen(text_alt) == 0) {
+                                g_free(text_alt);
+                                text_alt = g_strdup("Invisible Image");
```
Instead of using `Invisible Image` for everything let's fall back to our previous system, numbers.

`Notification with inline image [image #1]`
And in dmenu
`[#1] url`

bebehei replied:
I agree. Setting the text to "Invisible image" sounds rather mysterious than helpful.

bebehei replied:
Sorry, I've seen this just now.

> Notification with inline image [image #1]
> And in dmenu
> [#1] url

Please, no. As a user, I'd question the rationale removing it for links and then adding it again for images. As a user, I'd really insult this dev (myself).

Simply changing it to `[Hidden Image]` makes more sense IMO.

tsipinakis commented on test/markup.c:
```diff
@@ -45,12 +45,97 @@ TEST test_markup_transform(void)
+
+TEST test_markup_strip_img(void)
+{
+        RUN_TESTp(helper_markup_strip_img, "v <img> img",                                         "v  img",       NULL);
```
Can tests really be nested like that? I don't see anything indicating so in `greatest` docs.

bebehei replied:
Yes and no. The code is indicating it, but it's nowhere documented.

It works and it's no problem. In functionality it's far better than calling an extra function.

I wish, there would be an additional layer between test and assertion.

bebehei replied:
Ok a small background info on this: When an assertion fails, it immediately returns the current function (which should be a test).

As we test multiple strings (done via multiple assertions), a failed assertion breaks the test function. By calling another subtest, the supertest does not fail and will test further for all strings.

bebehei commented:
@tsipinakis I've rebased the fixes already into their commits. If you want to see the actual diff, please see here: https://github.com/bebehei/dunst/compare/refactor-notification_init...refactor-notification_init-backup

bebehei commented:
Finally!
